File: web/src/lib/stores/apps.ts
import { get, writable } from 'svelte/store'
import { live } from '../live/ws'
import { api } from '../api/client'

export interface App {
  id: string
  name: string
  icon: string
  status: 'running' | 'stopped' | 'partial'
  managed: boolean
  store?: string
  scheme?: string
  hostname?: string
  port?: string
  index?: string
  category?: string
  /** Fully-resolved click URL from x-compose-app (webui-*); wins when set. */
  url?: string
  /** Aggregated Docker health-check verdict; drives the tile status dot.
   *  Absent when no container declares a health check. */
  health?: 'healthy' | 'unhealthy' | 'starting'
  /** Pinned by the operator (PROTECTED_APPS): the tile hides its Uninstall
   *  entry, and the backend refuses a DELETE for it. */
  protected?: boolean
  /** True while a lifecycle op (start/stop/restart/uninstall) is in flight:
   *  the tile shows a "…" overlay and hides its burger menu. */
  busy?: boolean
  /** True while a store install is in flight for this app: the tile shows two
   *  progress bars (download + start) instead of being clickable. */
  installing?: boolean
  /** Image-pull progress, 0-100 (only meaningful while installing). */
  download?: number
  /** Stack-start progress, 0-100, driven by Docker (only while installing). */
  start?: number
  /** Current install phase: pull | prepare | start | done | error. */
  phase?: string
  /** Set when an install failed; the tile shows the error until retried. */
  install_error?: string
}

// Last-known app list, cached in localStorage so a page reload paints the grid
// instantly — before the backend (or Docker) has answered. The server is now the
// source of truth for *what exists* (folder-driven, docs/app-model.md), so the
// cache only bridges the first-paint gap; a fresh snapshot supersedes it in ms.
const CACHE_KEY = 'casadash.apps'

/** Read the cached tiles, or [] when absent/corrupt. */
function cachedApps(): App[] {
  try {
    const v = JSON.parse(localStorage.getItem(CACHE_KEY) ?? '[]')
    return Array.isArray(v) ? (v as App[]) : []
  } catch {
    return []
  }
}

/** Persist tiles for the next reload, minus volatile in-flight overlays — so a
 *  reload mid-install/-op never resurrects a stuck "…" or a frozen progress bar.
 *  Identity, name, icon and last-known status/health are kept so the tile looks
 *  right immediately (greyed vs coloured) until the live state refreshes. */
function persistApps(list: App[]): void {
  try {
    const clean = list.map((a) => {
      const c: App = { ...a }
      delete c.busy
      delete c.installing
      delete c.download
      delete c.start
      delete c.phase
      delete c.install_error
      return c
    })
    localStorage.setItem(CACHE_KEY, JSON.stringify(clean))
  } catch {
    /* private mode / quota — cache is best-effort */
  }
}

/** Apply an authoritative snapshot: replace the grid and refresh the cache. */
function applySnapshot(list: App[]): void {
  apps.set(list)
  persistApps(list)
}

export const apps = writable<App[]>(cachedApps())

/** One-shot REST load (used on mount for an immediate render). The REST endpoint
 *  is authoritative — including a genuine empty list — so its result always wins;
 *  on failure we keep whatever is on screen (cache or a prior load). */
export async function loadApps(): Promise<void> {
  try {
    applySnapshot((await api.get<App[]>('/api/apps')) ?? [])
  } catch {
    /* leave previous value (cached tiles stay visible) */
  }
}

/** Live updates over the WebSocket "apps" channel. Returns unsubscribe.
 *
 *  Non-destructive: an empty live frame never blanks a populated grid. Post
 *  step-1 the server only emits [] when there genuinely are no apps or it is
 *  still warming up, and those cases are already covered by loadApps()'s
 *  authoritative REST replace (mount + after every mutation, incl. uninstall).
 *  Ignoring empty frames here trades a rare stale ghost tile — self-healing on
 *  the next reload/mutation — for never flashing an empty dashboard. */
export function subscribeApps(): () => void {
  return live.subscribe('apps', (d) => {
    const list = (d as App[]) ?? []
    if (list.length === 0 && get(apps).length > 0) return
    applySnapshot(list)
  })
}

export async function appAction(id: string, action: 'start' | 'stop' | 'restart'): Promise<void> {
  await api.post(`/api/apps/${encodeURIComponent(id)}/${action}`)
  await loadApps()
}

/** Uninstall an app. Its folder is always preserved — renamed to
 *  `<app>.<date>.archive` (never deleted). When zip is true it is compressed to
 *  a `.zip` archive instead of a plain rename. Returns the archive's name. */
export async function uninstallApp(id: string, zip = false): Promise<string> {
  const res = await api.del<{ status: string; archive?: string }>(
    `/api/apps/${encodeURIComponent(id)}?zip=${zip}`,
  )
  await loadApps()
  return res?.archive ?? ''
}

/** Effective opening-URL (x-compose-app webui-*) config for an app, plus the
 *  server-resolved preview URL. Editing it writes into the app's override. */
export interface WebUI {
  scheme: string
  host: string
  port: string
  path: string
  url: string
}

export interface AppConfig {
  base: string
  override: string
  webui: WebUI
  /** App guidance — seeded from the store, editable; saved into the override. */
  tips: string
}

export function getConfig(id: string): Promise<AppConfig> {
  return api.get<AppConfig>(`/api/apps/${encodeURIComponent(id)}/config`)
}

/** Save (or clear, when blank) the app's tips into its override. */
export async function setTips(id: string, tips: string): Promise<void> {
  await api.put(`/api/apps/${encodeURIComponent(id)}/tips`, { tips })
}

/** Fetch the app's tips with ${VAR} references resolved (env-replaced preview). */
export async function renderTips(id: string): Promise<string> {
  const r = await api.get<{ tips: string }>(`/api/apps/${encodeURIComponent(id)}/tips`)
  return r.tips
}

export async function setConfig(id: string, override: string): Promise<void> {
  await api.put(`/api/apps/${encodeURIComponent(id)}/config`, { override })
  await loadApps()
}

/** Save the opening-URL fields into the app's override (webui-* shortcut). */
export async function setWebUI(id: string, w: Omit<WebUI, 'url'>): Promise<void> {
  await api.put(`/api/apps/${encodeURIComponent(id)}/webui`, w)
  await loadApps()
}

/** The app's .env, in file order — prefilled at install (PUID, DATA_ROOT, REF_*,
 *  …) and hand-editable since. It is what ${VAR} in the app's compose resolves
 *  against; it is NOT the same as a service's `environment:` block (that lives in
 *  the override form). Shares the EnvVar shape declared below. */
export function getEnv(id: string): Promise<EnvVar[]> {
  return api.get<EnvVar[]>(`/api/apps/${encodeURIComponent(id)}/env`)
}

/** Rewrite the app's .env to exactly these vars and recreate the stack — compose
 *  reads .env only when it brings the project up. */
export async function setEnv(id: string, vars: EnvVar[]): Promise<void> {
  await api.put(`/api/apps/${encodeURIComponent(id)}/env`, { vars })
  await loadApps()
}

/** A single-valued override field. `value` is what runs, `base` what the store
 *  ships; clearing `value` resets the field to the store's. `complex` marks a
 *  construct only the YAML view can edit safely (a list-form command, a hand-
 *  written tag) — the form shows it read-only. */
export interface Scalar {
  value: string
  base: string
  overridden: boolean
  complex: boolean
  /** The field's YAML, when `complex` — so the form can show what it won't edit. */
  raw?: string
}

/** A sequence field (ports, volumes, …). `value` is the effective list after
 *  Compose's merge; entries also present in `base` came from the store. */
export interface ListField {
  value: string[]
  base: string[]
  overridden: boolean
  complex: boolean
  raw?: string
}

export interface EnvVar {
  key: string
  value: string
}

export interface EnvField {
  value: EnvVar[]
  base: EnvVar[]
  overridden: boolean
  complex: boolean
  raw?: string
}

/** One compose service as the settings form edits it. */
export interface FormService {
  name: string
  image: Scalar
  restart: Scalar
  ports: ListField
  volumes: ListField
  environment: EnvField
  privileged: Scalar
  command: Scalar
  mem_limit: Scalar
  cpus: Scalar
  devices: ListField
  cap_add: ListField
}

export interface OverrideForm {
  services: FormService[]
}

/** The friendly view of the app's override — one entry per compose service. */
export function getOverrideForm(id: string): Promise<OverrideForm> {
  return api.get<OverrideForm>(`/api/apps/${encodeURIComponent(id)}/override/form`)
}

/** Patch the override with the form's values and recreate the app. */
export async function setOverrideForm(id: string, form: OverrideForm): Promise<void> {
  await api.put(`/api/apps/${encodeURIComponent(id)}/override/form`, form)
  await loadApps()
}

/** Check a candidate override against the app's base compose without applying it.
 *  A rejected override is a normal response carrying Compose's own message. */
export function validateOverride(id: string, override: string): Promise<{ ok: boolean; error?: string }> {
  return api.post<{ ok: boolean; error?: string }>(
    `/api/apps/${encodeURIComponent(id)}/override/validate`,
    { override },
  )
}

/** The project as Compose resolves it: base + override, merged and interpolated. */
export async function effectiveConfig(id: string): Promise<string> {
  const r = await api.get<{ config?: string; error?: string }>(
    `/api/apps/${encodeURIComponent(id)}/effective`,
  )
  if (r.error) throw new Error(r.error)
  return r.config ?? ''
}

/** Whether an app's reference store carries a newer docker-compose.yml than the
 *  installed copy. Resolved from the store reference recorded in the override. */
export interface UpdateStatus {
  /** The app records where it was installed from (needed to check at all). */
  has_ref: boolean
  /** The store's compose differs from the installed one. */
  available: boolean
  /** Reference store URL and the catalog id within it. */
  store: string
  store_app_id: string
  /** Non-fatal lookup failure (store unreachable, app removed from catalog, …). */
  error?: string
}

/** Check whether the app's reference store has a newer compose than installed. */
export function checkUpdate(id: string): Promise<UpdateStatus> {
  return api.get<UpdateStatus>(`/api/apps/${encodeURIComponent(id)}/update`)
}

/** Pull the store's current compose (when it differs) and bring the stack back
 *  up. Returns true when an update was actually applied. */
export async function applyUpdate(id: string): Promise<boolean> {
  const res = await api.post<{ status: string; updated: boolean }>(
    `/api/apps/${encodeURIComponent(id)}/update`,
  )
  await loadApps()
  return res?.updated ?? false
}

/** One container of a multi-service stack, with live state and health. */
export interface AppService {
  service: string
  container_id: string
  state: string
  health: string // '', starting, healthy, unhealthy
}

export function getServices(id: string): Promise<AppService[]> {
  return api.get<AppService[]>(`/api/apps/${encodeURIComponent(id)}/services`)
}

/** Compute an app's web URL, if it has a reachable one. */
export function appUrl(a: App): string {
  // x-compose-app (webui-*) resolves the full URL server-side; use it verbatim.
  if (a.url) return a.url
  const scheme = a.scheme || 'http'
  const index = a.index && a.index !== '/' ? a.index : ''
  if (a.hostname) return `${scheme}://${a.hostname}${index}`
  if (a.port) return `${scheme}://${location.hostname}:${a.port}${index}`
  return ''
}

/** Open an app, or warn if it has no reachable URL.
 *
 *  Optimistic routing (matching CasaOS): an app the tile already shows as cleanly
 *  running goes STRAIGHT to its URL — no interstitial, so the nominal case has no
 *  flash. Only an app that isn't cleanly up (stopped, partial, still health-check
 *  "starting", or unhealthy) goes through `/launch?app=<id>`, which starts the
 *  stack if needed and holds a friendly status page until it responds — instead of
 *  a browser connection error or a gateway 502. Both paths open synchronously in
 *  the click handler so the popup blocker stays clear. */
export function openApp(a: App): void {
  const url = appUrl(a)
  if (url === '') {
    alert(
      `${a.name} has no directly reachable web address.\n\n` +
        `It exposes its UI to a reverse-proxy/gateway rather than a host port. ` +
        `Add a published port via the app's Settings → override, or put it behind a gateway.`,
    )
    return
  }
  const cleanlyUp = a.status === 'running' && a.health !== 'starting' && a.health !== 'unhealthy'
  const target = cleanlyUp ? url : `/launch?app=${encodeURIComponent(a.id)}`
  window.open(target, '_blank', 'noopener')
}

File: internal/server/apps.go
package server

import (
	"context"
	"net/http"
	"time"

	"github.com/go-chi/chi/v5"

	"github.com/yundera/casadash/internal/apps"
)

func (s *Server) requireApps(w http.ResponseWriter) bool {
	if s.apps == nil {
		writeJSON(w, http.StatusServiceUnavailable,
			map[string]string{"error": "docker unavailable"})
		return false
	}
	return true
}

func (s *Server) handleListApps(w http.ResponseWriter, r *http.Request) {
	if !s.requireApps(w) {
		return
	}
	list := s.listApps(r.Context())
	if list == nil {
		list = []apps.App{}
	}
	writeJSON(w, http.StatusOK, list)
}

func (s *Server) handleAppAction(w http.ResponseWriter, r *http.Request) {
	if !s.requireApps(w) {
		return
	}
	id := chi.URLParam(r, "id")
	action := chi.URLParam(r, "action")

	var err error
	switch action {
	case "start":
		err = s.apps.Start(r.Context(), id)
	case "stop":
		err = s.apps.Stop(r.Context(), id)
	case "restart":
		err = s.apps.Restart(r.Context(), id)
	default:
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "unknown action"})
		return
	}
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
		return
	}
	writeJSON(w, http.StatusOK, map[string]string{"status": "ok"})
}

// handleCheckUpdate reports whether the app's reference store carries a newer
// docker-compose.yml than the installed copy (see installer.CheckUpdate).
func (s *Server) handleCheckUpdate(w http.ResponseWriter, r *http.Request) {
	if !s.requireApps(w) {
		return
	}
	if s.installer == nil {
		writeJSON(w, http.StatusServiceUnavailable, map[string]string{"error": "store unavailable"})
		return
	}
	ctx, cancel := context.WithTimeout(r.Context(), 90*time.Second)
	defer cancel()
	st, err := s.installer.CheckUpdate(ctx, chi.URLParam(r, "id"))
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
		return
	}
	writeJSON(w, http.StatusOK, st)
}

// handleApplyUpdate copies the store's current compose over the app's strict base
// (when it differs) and brings the stack back up. The tile shows a "…" overlay
// while it runs.
func (s *Server) handleApplyUpdate(w http.ResponseWriter, r *http.Request) {
	if !s.requireApps(w) {
		return
	}
	if s.installer == nil {
		writeJSON(w, http.StatusServiceUnavailable, map[string]string{"error": "store unavailable"})
		return
	}
	id := chi.URLParam(r, "id")
	ctx, cancel := context.WithTimeout(r.Context(), 5*time.Minute)
	defer cancel()

	var updated bool
	err := s.apps.WithBusy(id, func() error {
		var e error
		updated, e = s.installer.ApplyUpdate(ctx, id)
		return e
	})
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
		return
	}
	s.broadcastApps()
	writeJSON(w, http.StatusOK, map[string]any{"status": "ok", "updated": updated})
}

func (s *Server) handleUninstallApp(w http.ResponseWriter, r *http.Request) {
	if !s.requireApps(w) {
		return
	}
	id := chi.URLParam(r, "id")
	// zip=true compresses the archived folder; otherwise it is a plain rename.
	// Either way the app's data is preserved (docs/app-model.md).
	zip := r.URL.Query().Get("zip") == "true"
	archiveName, err := s.apps.Uninstall(r.Context(), id, zip)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
		return
	}
	// Drop any lingering install-progress overlay (e.g. a failed install) so the
	// tile disappears with the archived app instead of ghosting as an error.
	if s.installer != nil {
		s.installer.ClearInstall(id)
	}
	s.broadcastApps()
	writeJSON(w, http.StatusOK, map[string]string{"status": "ok", "archive": archiveName})
}
